### PR TITLE
Use context managers to simplify test_sanity

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

File: tests/test_sanity.py [hunk 1]
```diff
@@ -745,12 +733,6 @@ def test():
           assert 'js optimizer using native' in output
           test() # still works
 
-      finally:
```
Comment (jgravelle-google) [inline tests/test_sanity.py]: I was going to complain about adding a dependency but removing these try/finallys is so dang nice

File: tests/test_sanity.py [hunk 2]
```diff
@@ -584,13 +576,12 @@ def test_emconfig(self):
...
-    self.do([PYTHON, EMCC, '--em-config', custom_config_filename] + MINIMAL_HELLO_WORLD + ['-O2'])
-    result = run_js('a.out.js')
-    
+    with chdir(temp_dir):
+      self.do([PYTHON, EMCC, '--em-config', custom_config_filename] + MINIMAL_HELLO_WORLD + ['-O2'])
+      result = run_js('a.out.js')
+
     # Clean up created temp files.
     os.remove(custom_config_filename)
-    os.chdir(path_from_root())
```
Comment (jgravelle-google) [inline tests/test_sanity.py]: This one just wasn't even trying to clean up was it
So if the emcc invocation failed, would the chdir(temp_dir) be persisted across tests, or is there some other mechanism to keep things from getting that out of hand?
Either way this is better

File: tests/test_sanity.py [hunk 3]
```diff
@@ -584,13 +576,12 @@ def test_emconfig(self):
...
-    os.chdir(temp_dir)
-    self.do([PYTHON, EMCC, '--em-config', custom_config_filename] + MINIMAL_HELLO_WORLD + ['-O2'])
-    result = run_js('a.out.js')
-    
+    with chdir(temp_dir):
+      self.do([PYTHON, EMCC, '--em-config', custom_config_filename] + MINIMAL_HELLO_WORLD + ['-O2'])
+      result = run_js('a.out.js')
+
     # Clean up created temp files.
     os.remove(custom_config_filename)
```
Comment (jgravelle-google) [inline tests/test_sanity.py]: Should this be scoped to the chdir(temp_dir), or is the custom_config_filename an absolute path?

File: tests/test_sanity.py [hunk 4]
```diff
@@ -33,10 +37,21 @@ def add_to_config(content):
...
   return os.stat(filename).st_mtime
 
+
+@contextlib.contextmanager
+def chdir(dir):
+  """A context manager that performs actions in the given directory."""
+  orig_cwd = os.getcwd()
+  os.chdir(dir)
+  try:
+    yield
```
Comment (jgravelle-google) [inline tests/test_sanity.py]: Didn't know this worked, cool

File: tests/test_sanity.py [hunk 5]
```diff
@@ -33,10 +37,21 @@ def add_to_config(content):
   with open(CONFIG_FILE, 'a') as f:
     f.write(content + '\n')
 
-
 def mtime(filename):
   return os.stat(filename).st_mtime
 
+
+@contextlib.contextmanager
+def chdir(dir):
```
Comment (jgravelle-google) [inline tests/test_sanity.py]: Talked in-person, probably better to not add a dependency on `mock`, but to write another wrapper for `patchEnv` or so

File: tests/test_sanity.py (hunk 3 above)
Comment (sbc100) [reply tests/test_sanity.py]: No its absolute.

File: tests/test_sanity.py (hunk 5 above)
Comment (sbc100) [reply tests/test_sanity.py]: Yeah, I think that will be cleaner.

File: tests/test_sanity.py [hunk 6]
```diff
@@ -1,4 +1,6 @@
 from __future__ import print_function
+import contextlib
+import mock
```
Comment (jgravelle-google) [inline tests/test_sanity.py]: Remove mock import

**Pull request merged**